group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,393 @@
+    // Indicates if a rack aware assignment can be done.
+    // True if racks are defined for both members and partitions.
+    boolean useRackAwareStrategy;
+
+    OptimizedUniformAssignmentBuilder(AssignmentSpec assignmentSpec, SubscribedTopicDescriber subscribedTopicDescriber) {
+        this.assignmentSpec = assignmentSpec;
+        this.subscribedTopicDescriber = subscribedTopicDescriber;
+        this.subscriptionIds = new HashSet<>(assignmentSpec.members().values().iterator().next().subscribedTopicIds());
+        this.rackInfo = new RackInfo(assignmentSpec, subscribedTopicDescriber, subscriptionIds);
+        this.potentiallyUnfilledMembers = new HashMap<>();
+        this.unfilledMembers = new HashMap<>();
+        this.newAssignment = new HashMap<>();
+        this.useRackAwareStrategy = rackInfo.useRackStrategy;
+        // Without rack-aware strategy, tracking current owners of unassigned partitions is unnecessary
+        // as all sticky partitions are retained until a member meets its quota.
+        this.currentPartitionOwners = useRackAwareStrategy ? new HashMap<>() : Collections.emptyMap();
+    }
+
+    /**
+     * Here's the step-by-step breakdown of the assignment process:
+     *
+     * <li> Compute the quotas of partitions for each member based on the total partitions and member count.</li>
+     * <li> For existing assignments, retain partitions based on the determined quota and member's rack compatibility.</li>
+     * <li> If a partition's rack mismatches with its owner, track it for future use.</li>
+     * <li> Identify members that haven't fulfilled their partition quota or are eligible to receive extra partitions.</li>
+     * <li> Derive the unassigned partitions by taking the difference between total partitions and the sticky assignments.</li>
+     * <li> Depending on members needing extra partitions, select members from the potentially unfilled list
+     *      and add them to the unfilled list.</li>
+     * <li> Proceed with a round-robin assignment adhering to rack awareness.
+     *      For each unassigned partition, locate the first compatible member from the unfilled list.</li>
+     * <li> If no rack-compatible member is found, revert to the tracked current owner.
+     *      If that member can't accommodate the partition due to quota limits, resort to a generic round-robin assignment.</li>
+     * */
+    @Override
+    protected GroupAssignment buildAssignment() throws PartitionAssignorException{
+        int totalPartitionsCount = 0;
+        // Removes the current topic from subscriptionIds if the topic doesn't exist in the topic metadata.
+        for (Uuid topicId : subscriptionIds) {
+            int partitionCount = subscribedTopicDescriber.numPartitions(topicId);
+            if (partitionCount == -1) {
+                IllegalStateException illegalStateException =
+                    new IllegalStateException("The topic " + topicId + " seems to be in an invalid state.");
```

Comment:
I wanted to throw a partition assignor exception since the exception is related to the assignor? can we do that only? Or do we have to do just an illegal state exception?

Reply:
Discussed offline, going to throw only PartitionAssignorException